File: 155-min-stack/min-stack.ts
class MinStack {
    private mainStack: number[]
    private minStack: number[]

    constructor() {
        this.mainStack = []
        this.minStack = []
    }

    push(val: number): void {
        this.mainStack.push(val)

        if (this.minStack.length === 0 || val <= this.getMin()) {
            this.minStack.push(val)
        }
    }

    pop(): void {
        if (this.mainStack.length === 0) {
            throw new Error('Stack is empty')
        }

        const popped = this.mainStack.pop()
        if (popped === this.getMin()) {
            this.minStack.pop()
        }
    }

    top(): number {
        if (this.mainStack.length === 0) {
            throw new Error('Stack is empty')
        }
        return this.mainStack[this.mainStack.length - 1]
    }

    getMin(): number {
        if (this.minStack.length === 0) {
            throw new Error('Stack is empty')
        }
        return this.minStack[this.minStack.length - 1]
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(val)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */

File: 155-min-stack/min-stack.py
class MinStack:

    def __init__(self):
        self.vals = []
        self.mins = []

    def push(self, val: int) -> None:
        self.vals.append(val)
        if not self.mins or val < self.mins[-1]:
            self.mins.append(val)

        else:
            self.mins.append(self.mins[-1])

    def pop(self) -> None:
        self.vals.pop()
        self.mins.pop()

    def top(self) -> int:
        return self.vals[-1]

    def getMin(self) -> int:
        return self.mins[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()